pennylane/transforms/sign_expand.py

```diff
@@ -0,0 +1,327 @@
+    """
+    mat = qml.utils.sparse_hamiltonian(hamiltonian).toarray()
+    size = len(mat)
+    eigs, eigvecs = np.linalg.eigh(mat)
+    norm = eigs[-1]
+    proj = np.identity(size, dtype="complex64")
+    proj += -2 * np.outer(np.conjugate(eigvecs[:, 0]), eigvecs[:, 0])
+    last_i = 1
+
+    dEs, mus, projs, times = [], [], [], []
+
+    for index in range(len(eigs) - 1):
+        dE = (eigs[index + 1] - eigs[index]) / 2
+        if np.isclose(dE, 0):
+            continue
+        dEs.append(dE)
+        mu = (eigs[index + 1] + eigs[index]) / 2
+        mus.append(mu)
+        time = np.pi / (2 * (norm + abs(mu)))
+        times.append(time)
+
+        for j in range(last_i, index + 1):
+            proj += -2 * np.outer(np.conjugate(eigvecs[:, j]), eigvecs[:, j])
+            last_i = index + 1
+
+        projs.append(proj.copy() * dE)
+
+    return dEs, mus, times, projs
+
+
+# pylint: disable=too-many-function-args)
+def construct_sgn_circuit(hamiltonian, tape, mus, times, phis):
+    """
+    Takes a tape with state prep and ansatz and constructs the individual tapes approximating/estimating the individual terms of your decomposition
+
+    Args:
+      hamiltonian (qml.Hamiltonian): The pennylane hamiltonian to be decomposed
+      tape (qml.QuantumTape: Tape containing the circuit to be expanded into the new circuits
+      mus (List[float]): The average between the two eigenvalues (E_1+E-2)/2
+      times (List[float]): The time for this term group to be evaluated/evolved at
+      phis (List[float]): Optimal phi values for the QSP part associated with the respective delta and J
+
+    Returns:
+      tapes (List[qml.tape]): Expanded tapes from the original tape that measures the terms via the approximate sgn decomposition
+    """
+    coeffs = hamiltonian.data
+    tapes = []
+    for mu, time in zip(mus, times):
+        with tape.__class__() as new_tape:
+            # Put state prep and ansatz on tape in the 'old' register
+            for op in tape.operations:
+                op.queue()
+
+            # Put QSP and Hadamard test on the two ancillas Target and Control
+            qml.Hadamard("Hadamard")
+            for i, phi in enumerate(phis):
+                qml.CRX(phi, wires=["Hadamard", "Target"])
+                if i == len(phis) - 1:
+                    qml.CRY(np.pi, wires=["Hadamard", "Target"])
+                else:
+                    evolve_under(hamiltonian.ops, coeffs, 2 * time)
+                    qml.CRZ(-2 * mu * time, wires=["Hadamard", "Target"])
+            qml.Hadamard("Hadamard")
+
+            if tape.measurements[0].return_type == qml.measurements.Expectation:
+                qml.expval(-1 * qml.PauliZ("Hadamard"))
+            else:
+                qml.var(qml.PauliZ("Hadamard"))
+
+        tapes.append(new_tape)
+    return tapes
+
+
+def sign_expand(tape, circuit=False, J=10, delta=0.0):
```

Comment:
You need to add this decorator such that this transform can be applied to quantum circuits.
```suggestion
@qml.transforms.batch_transform
def sign_expand(tape, circuit=False, J=10, delta=0.0):
```
Example:
```python
dev = qml.device("default.qubit", wires=[0, 1, 2, 3, "Hadamard", "Target"])
H4 = (
    qml.PauliX(0) @ qml.PauliZ(2)
    + 3 * qml.PauliZ(2)
    - 2 * qml.PauliX(0)
    + qml.PauliZ(2)
    + qml.PauliZ(2)
)
@qml.transforms.sign_expand
@qml.qnode(dev)
def circuit():
    qml.Hadamard(0)
    qml.Hadamard(1)
    qml.PauliZ(1)
    qml.PauliX(2)
    return qml.expval(H4)
```
```python
>>> circuit()
-6.999999760400206
```
**Note:** The `processing_fn` is executed under the hood. Please check [the docs](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.batch_transform.html?highlight=batch%20transform#qml-batch-transform) for more examples.
Also please add/change some tests to use quantum functions instead of tapes.

Reply:
I think for tests we can use either/ both quantum functions and tapes, but for the documentation, we should use quantum functions and qnodes.  We are trying to hide tapes away from users.

Reply:
Added the decorator, your suggestion leads to a circular import so I'm importing the function manually from .batch_transform to avoid that